lib/preboots/scheduler.js

```diff
@@ -0,0 +1,234 @@
+    async.eachLimit(packages, this.conc, (pkg, next) => {
+      const fn = once(next);
+      const name = pkg.name;
+      this.models.BuildHead.findAll({ name, env }, (err, heads) => {
+        if (err) return fn(err);
+        if (!heads || !heads.length) return fn();
+
+        const lookup = new Lookup(pkg, heads);
+        const writer = nsqStream.createWriteStream(this.nsq.writer, this.topic);
+
+        const missing = lookup.missing();
+        const len = missing.length;
+        this.log.info('%d missing builds. triggering new builds for %s', len, name);
+        counts[name] = len;
+
+        from.obj(missing)
+          .pipe(writer)
+          .on('error', fn)
+          .once('finish', fn);
+      });
+    }, done);
+  });
+};
+
+/**
+ * Fetch all packages we currently have stored using our special cache table
+ * for fast fetches
+ * @function packages
+ * @api public
+ * @param {Function} fn - Continuation function to call when completed
+ * @returns {undefined}
+ */
+Scheduler.prototype.packages = function packs(fn) {
+  this.models.PackageCache.findAll({ partitioner: 'cached' }, fn);
+};
+
+/**
+ * Lookup Object for parsing build heads
+ * @constructor
+ * @param {Object} pkg - package object
+ * @param {Object} heads - build heads for a given package in the fetched env
+ */
+function Lookup(pkg, heads) {
+  this.pkg = pkg;
+  this.heads = heads;
+  this.latest = this.order(heads);
+}
+
+/**
+ * Compute and order the heads to return the latest version
+ * @function order
+ * @param {Array} heads - Array of head objects
+ * @returns {String} latest version of the set
+ */
+Lookup.prototype.order = function order(heads) {
+  return heads.filter(Boolean)
+    .sort((a, b) => {
+      return semver.lt(a.version, b.version) ? -1 : 1;
+    }).reduce((latest, v) => {
+      return semver.lt(latest, v.version)
+        ? v.version
+        : latest;
+    }, '0.0.0');
+};
+
+/**
+ * Compute objects to be sent over nsq for missing builds
+ * @function missing
+ * @returns {Array} of spec objects to be sent to nsq
+ */
+Lookup.prototype.missing = function () {
+  return this.heads.map((head) => {
+    return semver.lt(head.version, this.latest)
+      ? this.specify(head)
+      : null;
+  }).filter(Boolean);
+};
+
+/**
+ * Create a spec object given a head object and internal data
+ * @function specify
+ * @param {Object} head - A given head object from a missing build
+ * @returns {Spec} object to be sent to nsq
+ */
+Lookup.prototype.specify = function specify(head) {
+  return new Spec(
+    head.name,
+    head.env,
+    this.latest,
+    head.locale,
+    this.pkg.extended.build
+  );
+};
+
+/**
+ * Spec Object
+ * @constructor
+ * @param {String} name - name of package
+ * @param {String} env - env of package
+ * @param {String} version - version of package
+ * @param {String} locale - locale of package
+ * @param {String} type - build type
+ */
+function Spec(name, env, version, locale, type) {
```

Comment:
any particular reason this cannot be a regular object?

Reply:
yea it could, this is just taking advantage of that old v8 optimization about preventing unamed virtual hidden classes (or something similar. It might not be as meaningful as it once was but thought it couldnt hurt